x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java

```diff
@@ -159,11 +159,15 @@ public boolean hasFieldOrDocumentLevelSecurity() {
             }
         }
         final StringMatcher nameMatcher = indexMatcher(ordinaryIndices, restrictedIndices);
-        final StringMatcher bwcSpecialCaseMatcher = indexMatcher(grantMappingUpdatesOnIndices, grantMappingUpdatesOnRestrictedIndices);
-        return indexAbstraction -> nameMatcher.test(indexAbstraction.getName())
-            || (indexAbstraction.getType() != IndexAbstraction.Type.DATA_STREAM
-                && (indexAbstraction.getParentDataStream() == null)
-                && bwcSpecialCaseMatcher.test(indexAbstraction.getName()));
+        if (grantMappingUpdatesOnIndices.isEmpty() && grantMappingUpdatesOnRestrictedIndices.isEmpty()) {
+            return indexAbstraction -> nameMatcher.test(indexAbstraction.getName());
+        } else {
+            final StringMatcher bwcSpecialCaseMatcher = indexMatcher(grantMappingUpdatesOnIndices, grantMappingUpdatesOnRestrictedIndices);
+            return indexAbstraction -> nameMatcher.test(indexAbstraction.getName())
+                || (indexAbstraction.getType() != IndexAbstraction.Type.DATA_STREAM
+                    && (indexAbstraction.getParentDataStream() == null)
+                    && bwcSpecialCaseMatcher.test(indexAbstraction.getName()));
+        }
```

Comment:
Just realised that this block of changes is not related to the purpose of this PR. It's a minor optimisation for `allowedIndicesMatcher`

Reply:
I think these sort of small things are a good thing, I don't think they harm the readability.

Reply:
It's not about readability. It is a very minor optimisation to not create `bwcSpecialCaseMatcher` and subsequently using it if we know it's not necessary. That said, the change has nothing to do with this PR, it's now reverted.